source/AST/ASTVisitor.cpp

```diff
@@ -1582,58 +1461,251 @@ TraverseFunctionTemplateDecl(
+            std::forward<Args>(args)...);
+        break;
+    case Decl::ClassTemplatePartialSpecialization:
+        this->Traverse(static_cast<
+            ClassTemplatePartialSpecializationDecl*>(D),
+            std::forward<Args>(args)...);
+        break;
+    case Decl::VarTemplate:
+        this->Traverse(static_cast<
+            VarTemplateDecl*>(D),
+            std::forward<Args>(args)...);
+        break;
+    case Decl::VarTemplateSpecialization:
+        this->Traverse(static_cast<
+            VarTemplateSpecializationDecl*>(D),
+            std::forward<Args>(args)...);
+        break;
+    case Decl::VarTemplatePartialSpecialization:
+        this->Traverse(static_cast<
+            VarTemplatePartialSpecializationDecl*>(D),
+            std::forward<Args>(args)...);
+        break;
+    case Decl::FunctionTemplate:
+        this->Traverse(static_cast<
+            FunctionTemplateDecl*>(D),
+            std::forward<Args>(args)...);
+        break;
+    case Decl::ClassScopeFunctionSpecialization:
+        this->Traverse(static_cast<
+            ClassScopeFunctionSpecializationDecl*>(D),
+            std::forward<Args>(args)...);
+        break;
+    case Decl::TypeAliasTemplate:
+        this->Traverse(static_cast<
+            TypeAliasTemplateDecl*>(D),
+            std::forward<Args>(args)...);
         break;
     default:
-        Assert(! "unhandled kind for FunctionDecl");
+        // for declarations we don't explicitly handle, traverse the children
+        // if it has any (e.g. LinkageSpecDecl, ExportDecl, ExternCContextDecl).
+        if(auto* DC = dyn_cast<DeclContext>(D))
+            TraverseContext(DC);
         break;
     }
-    // we don't care about any block scope declarations,
-    // nor do we need to traverse the FunctionDecl, so just return
+
     return true;
 }
 
 bool
 ASTVisitor::
-TraverseClassScopeFunctionSpecializationDecl(
-    ClassScopeFunctionSpecializationDecl* D)
+TraverseContext(
+    DeclContext* D)
 {
-    if(! shouldExtract(D))
-        return true;
+    Assert(D);
+    for(auto* C : D->decls())
+        TraverseDecl(C);
+    return true;
+}
 
-    FunctionInfo I;
-    parseTemplateArgs(I.Template, D);
+// An instance of Visitor runs on one translation unit.
+void
+ASTVisitor::
+HandleTranslationUnit(
+    ASTContext& Context)
+{
+    // cache contextual variables
+    astContext_ = &Context;
+    sourceManager_ = &astContext_->getSourceManager();
 
-    CXXMethodDecl* MD = D->getSpecialization();
-    buildFunction(I, MD);
+    // Install handlers for our custom commands
+    initCustomCommentCommands(Context);
 
-    return true;
+    std::optional<llvm::StringRef> filePath =
+        Context.getSourceManager().getNonBuiltinFilenameForID(
+            Context.getSourceManager().getMainFileID());
+    if(! filePath)
+        return;
+
+    // Filter out TUs we don't care about
+    File = *filePath;
+    convert_to_slash(File);
+    if(! config_.shouldVisitTU(File))
```

Comment:
broseph I do not think this is okay !

Reply:
Could you elaborate? I just moved this function -- this code was not changed by me

Reply:
Nevermind... I thought it was not thread-safe.